File: csharp/Prescribe.Core/Semantics/Symbols.cs
using Prescribe.Core.Frontend;

namespace Prescribe.Core.Semantics;

public enum SymbolKind
{
    Var,
    Const,
    Type,
    Proc,
    Func,
    Class,
    Field,
    Method,
    Constructor,
    Param,
    EnumMember
}

public sealed record Symbol(SymbolKind Kind, string Name, TypeSymbol? Type = null, AstNode? Decl = null, AccessModifier? Access = null, string? OwnerClass = null);

public sealed class Scope
{
    private readonly Dictionary<string, Symbol> _symbols = new(StringComparer.Ordinal);

    public Scope(Scope? parent = null)
    {
        Parent = parent;
    }

    public Scope? Parent { get; }

    public void Define(Symbol sym)
    {
        _symbols[sym.Name] = sym;
    }

    public Symbol? Lookup(string name)
    {
        if (_symbols.TryGetValue(name, out var sym)) return sym;
        return Parent?.Lookup(name);
    }

    public Symbol? LookupLocal(string name) => _symbols.TryGetValue(name, out var sym) ? sym : null;
}

public sealed record ClassInfo(
    string Name,
    string? BaseName,
    Dictionary<string, Symbol> Fields,
    Dictionary<string, Symbol> Methods,
    Symbol? Constructor,
    ClassDeclNode Decl);

public sealed record ProcInfo(string Name, AstNode Decl, TypeSymbol? Type = null);

public sealed record BlockInfo(Scope Scope, BlockNode Block);

File: legacy-typescript/src/semantics/symbols.ts
import { Type } from "./types.js";
import { BlockNode, ClassDeclNode, ProcDeclNode, FuncDeclNode, VarDeclNode, ConstDeclNode, TypeDeclNode, ConstructorDeclNode, ParamNode } from "../frontend/ast.js";

export type SymbolKind =
  | "Var"
  | "Const"
  | "Type"
  | "Proc"
  | "Func"
  | "Class"
  | "Field"
  | "Method"
  | "Constructor"
  | "Param"
  | "EnumMember";

export type Symbol = {
  kind: SymbolKind;
  name: string;
  type?: Type;
  decl?: VarDeclNode | ConstDeclNode | TypeDeclNode | ProcDeclNode | FuncDeclNode | ClassDeclNode | ConstructorDeclNode | ParamNode;
  access?: "PUBLIC" | "PRIVATE";
  ownerClass?: string;
};

export class Scope {
  private readonly symbols = new Map<string, Symbol>();
  constructor(public readonly parent?: Scope) {}

  define(sym: Symbol): void {
    this.symbols.set(sym.name, sym);
  }

  lookup(name: string): Symbol | undefined {
    if (this.symbols.has(name)) return this.symbols.get(name);
    return this.parent?.lookup(name);
  }

  lookupLocal(name: string): Symbol | undefined {
    return this.symbols.get(name);
  }
}

export type ClassInfo = {
  name: string;
  baseName?: string;
  fields: Map<string, Symbol>;
  methods: Map<string, Symbol>;
  ctor?: Symbol;
  decl: ClassDeclNode;
};

export type ProcInfo = {
  name: string;
  decl: ProcDeclNode | FuncDeclNode;
  type?: Type;
};

export type BlockInfo = {
  scope: Scope;
  block: BlockNode;
};
